app/models/seoable/seo_detail.rb

```diff
@@ -4,10 +4,21 @@ class SeoDetail < ActiveRecord::Base
     @@permitted_attribute_names = [:meta_title, :meta_description, :slug]
 
     belongs_to :seoable, polymorphic: true
+    has_many :redirects, class_name: 'SeoDetailRedirect'
+
+    after_update :create_redirect_if_needed
 
     validates :meta_title, presence: true, uniqueness: { scope: :seoable_type }
     validates :slug,
               uniqueness: { scope: :seoable_type },
               format: /\A[a-z0-9-]+\z/
+
+    scope :with_redirects, -> (model, slug){ includes(:redirects).where('seoable_type = ? AND (seo_detail_redirects.old_slug = ? OR slug = ?)', model, slug, slug).references(:redirects) }
```

Comment:
You could maybe store redirect info in the same table. Also check out how FriendlyId did it. They've been production tested for a long time on many apps.